src/thor/bidirectional_astar.cc

```diff
@@ -617,11 +643,31 @@ BidirectionalAStar::GetBestPath(valhalla::Location& origin,
       return {};
     }
 
+    // Exhaust hierarchy limits simultaneously in both directions. As soon as forward/reverse
+    // search exhausts limits on the particular level - it should stop and wait until reverse/forward
+    // search exhausts limits on the same hierarchy level. This logic ensures local optimality near
+    // the origin and destination and provides valid conditions for the reach-based pruning.
+    bool force_forward = false;
+    bool force_reverse = false;
+    assert(hierarchy_limits_forward_.size() == hierarchy_limits_forward_.size());
+    for (size_t level = TileHierarchy::levels().size() - 1; level > 0; --level) {
+      if (hierarchy_limits_reverse_[level].StopExpanding(rev_pred.distance()) &&
+          !hierarchy_limits_forward_[level].StopExpanding(fwd_pred.distance())) {
+        force_forward = true;
+        break;
+      } else if (hierarchy_limits_forward_[level].StopExpanding(fwd_pred.distance()) &&
+                 !hierarchy_limits_reverse_[level].StopExpanding(rev_pred.distance())) {
+        force_reverse = true;
```

Comment:
force reverse pass in case the forward search exhausted hierarchy limits for this level